components/messenger.rst

```diff
@@ -86,6 +86,46 @@ that will do the required processing for your message::
        }
     }
 
+Envelope
+--------
+
+The notion of an envelope is a concept that helps us to add some context around the
+messages. An envelope is a message and a set of items. On a user perspective, this
+allows to set some configuration around the message. For example the serialization
+groups used when the message goes through the transport layer::
+
+    use Symfony\Component\Messenger\Envelope;
+    use Symfony\Component\Messenger\Transport\Serialization\SerializerConfiguration;
+
+    $bus->dispatch(
+        (new Envelope($message))->with(new SerializerConfiguration([
+            'groups' => ['my_serialization_groups'],
+        ]))
+    );
+
+Instead of dealing directly with the messages in the handlers and middleware you
+can receive the envelope by implementing the ``EnvelopeAwareInterface`` marker
+interface on your middleware or handler like this::
+
+    use Symfony\Component\Messenger\Middleware\MiddlewareInterface;
+    use Symfony\Component\Messenger\EnvelopeAwareInterface;
+
+    class MyOwnMiddleware implements MiddlewareInterface, EnvelopeAwareInterface
+    {
+        public function handle($message, callable $next)
+        {
+            // $message here is an `Envelope` object
+
+            return $next(
+                $message->with(new AnotherEnvelopeItem(/* ... */))
+            );
```

Comment:
Is this the main use-case for an envelop-aware middleware? Or would we more commonly be reading some configuration from the envelope?

Reply:
Both are good use cases tbh